File: Week-4-April-22nd-28th/Jump-Game/Jump-game.py
# Time complexity: O(n) ... Space complexity: O(1)
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reach = 0
        for ind, val in enumerate(nums):
            if ind <= reach:
                reach = max(reach, ind + val)
                if reach >= len(nums) - 1:
                    return True
        return False

File: Week-4-April-22nd-28th/Jump-Game/Jump-game.cpp
// Time complexity: O(n) .... Space complexity: O(1)
class Solution {
public:
    bool canJump(vector<int>& nums) {
        if(nums.size() <= 1) return true;
        int ind = 0;
        // maxJump is the longest jump distance at a ith index
        for(int maxJump = 0; ind < nums.size() - 1 && ind <= maxJump; ++ind){
            maxJump = max(maxJump, ind + nums[ind]);
            if(maxJump >= nums.size() - 1){
                return true;
            }
        }
        return false;
    }
};